### PR TITLE
Add the ability for profilers to listen to EventPipe events

### LINKED ISSUE #11750: Support getting EventPipe events via ICorProfile callbacks
Historically we had an unnecessary divide between ETW events that were available asynchronously out-of-proc and ICorProfile API callbacks that were available synchronously in-proc. Some events were available via one source, some via the other, and some via both. Although it is technically possible to create profilers that register for both streams of events it is needlessly complex and it is impossible to make the async ETW events become synchronous again. To mitigate these issues we want to make all of the ETW events available synchronously via an ICorProfile callback. And because we are cross platform now, ETW events are better abstracted as EventPipe events so that this will work properly on non-Windows platforms too.

We already had a past effort at doing this in PR dotnet/coreclr#19157 and a little bit of discussion of it in dotnet/runtime#5638. The dotnet/coreclr#19157 attempt handled receiving the EventPipe events fine, but it didn't include a mechanism to enable the events. To use it, a profiler would need to configure the set of desired events via an out-of-band mechanism such as environment variables or managed EventListener APIs. Although this would work for some scenarios, I expected it would be onerous in the common case and it would be awkward to migrate from that form of API to one where ICorProfile also had direct control over the set of events it was receiving.

Work to do:

- [x] Add multi-session support to EventPipe so that the ICorProfile client's event settings need not interfere with other EventPipe clients. dotnet/runtime#9388, dotnet/runtime#11435
- [ ] Add EventPipe event control APIs to ICorProfile so that the profiler can select which events it wants to receive
- [ ] Update/complete any work necessary in the EventPipe listening APIs implemented in PR dotnet/coreclr#19157

### PR DESCRIPTION
**Pull request by davmason**
Fixes #11750

Marking this as draft because I'd like to review the overall design. This adds the APIs necessary for profilers to enable an EventPipe session and receive synchronous callbacks for the events.

I changed EventPipeSession to have one more mode, where it takes a callback and issues the callback synchronously instead of writing to the buffer manager. I purposefully did the callback higher up in the stack so that we don't waste any time encoding events and so that the profilers don't have to spend time/effort decoding them.

I have test code at https://github.com/davmason/eventpipe_profiler_test that I have tested this enough to be confident that it doesn't have any obvious bugs.

### CODE REVIEW

**Comment by davmason** [comment]
cc @mjsabby

File: src/coreclr/src/vm/eventpipesession.h
Comment (josalem) [inline src/coreclr/src/vm/eventpipesession.h:199]: Can we add a comment here that describes when this function is synchronous (under new profiler session type) and when it is asynchronous (all other times)?  The asynchronous nature of this method is important to the performance of the eventing pipeline and without `Buffered` in its name, it may not be immediately clear what the semantics are.

File: src/coreclr/src/vm/proftoeeinterfaceimpl.cpp
Comment (josalem) [inline src/coreclr/src/vm/proftoeeinterfaceimpl.cpp:7108]: nit: is there anything needed to rectify this comment?

File: src/coreclr/src/vm/proftoeeinterfaceimpl.cpp
Comment (josalem) [inline src/coreclr/src/vm/proftoeeinterfaceimpl.cpp:7170]: I don't think there is any locking/protection around adding providers to a session while the session is active.  There might be dragons here 🐉  if you use this while a session is alive.

File: src/coreclr/src/vm/eventpipesession.cpp
Comment (davmason) [reply src/coreclr/src/vm/eventpipesession.cpp:332]: I know it's a hot path, but I don't think we should be concerned about this particular if check. On the WriteEvent path we do lots of work that would dwarf this before we ever get here, like serializing data in EventSource. I also believe that the compiler and processor are pretty good about optimizing null checks.

I am planning on doing some perf testing before checking this in, if it shows up as a regression I'm happy to address it then

File: src/coreclr/src/vm/eventpipeconfiguration.cpp
Comment (davmason) [reply src/coreclr/src/vm/eventpipeconfiguration.cpp:164]: It skips the call if whatever is passed to the macro is false, so it will skip it even if a profiler is present but not listening to EventPipe events

File: src/coreclr/src/vm/eventpipesession.cpp
Comment (josalem) [reply src/coreclr/src/vm/eventpipesession.cpp:332]: Works for me 😃

File: src/coreclr/src/vm/proftoeeinterfaceimpl.cpp
Comment (davmason) [reply src/coreclr/src/vm/proftoeeinterfaceimpl.cpp:7170]: Thanks, I copied this from the rundown logic but missed that it was holding the lock. I will make sure to grab the lock.

File: src/coreclr/src/vm/proftoeeinterfaceimpl.cpp
Comment (mjsabby) [inline src/coreclr/src/vm/proftoeeinterfaceimpl.cpp:7293]: So we need the string allocation here? Any way to avoid it?

File: src/coreclr/src/vm/proftoeeinterfaceimpl.cpp
Comment (mjsabby) [reply src/coreclr/src/vm/proftoeeinterfaceimpl.cpp:7108]: I have the same question. The idea is the profiler api delivery is more efficient than the async path. Is that not correct?

File: src/coreclr/src/vm/proftoeeinterfaceimpl.cpp
Comment (davmason) [reply src/coreclr/src/vm/proftoeeinterfaceimpl.cpp:7293]: I don't think this actually does an allocation, we already have the provider name on the EventPipeProvider object. And I wouldn't expect this API to be on any hot paths, you would only call this if you want to figure out the name. You would generally use the handle to identify the provider you care about and only call this if you want to do some formatting. There's no reason you couldn't cache the result

File: src/coreclr/src/vm/proftoeeinterfaceimpl.cpp
Comment (davmason) [reply src/coreclr/src/vm/proftoeeinterfaceimpl.cpp]: It already was synchronous, but this comment made me realize we were still creating the buffer. I just pushed changes so now when using profiler we never even create the buffer.

File: src/coreclr/src/inc/corprof.idl
Comment (mjsabby) [inline src/coreclr/src/inc/corprof.idl:4118]: sessionName?

File: src/coreclr/src/inc/corprof.idl
Comment (mjsabby) [inline src/coreclr/src/inc/corprof.idl:4123]: requestTimerIntervalSamples or requestCpuSamples

File: src/coreclr/src/inc/corprof.idl
Comment (mjsabby) [inline src/coreclr/src/inc/corprof.idl:4122]: requestMetadataDump?

File: src/coreclr/src/vm/eventpipe.cpp
Comment (davmason) [reply src/coreclr/src/vm/eventpipe.cpp:325]: After looking at the code more, rundownRequested signals whether you want rundown to happen when the session closes, rundownEnabled sets the `m_rundownEnabled` flag in EventpipeSession.

But, it had a default argument of false in the constructor and no one ever calls it with a different value. Looking at the logic in EventPipeSession::EnableRundown it wouldn't even cause rundown to happen if specified. I assume it was an oversight from the original code, or used to be used and all code paths have been refactored so it's dead code.

I removed it and it didn't cause any compilation errors so I'll just remove it as part of this change

**Review by mjsabby — APPROVED** [review]

**Review by noahfalk — COMMENTED** [review]
Looks like a good start! Various comments inline

File: src/coreclr/src/inc/corprof.idl
Comment (davmason) [reply src/coreclr/src/inc/corprof.idl:650]: It's only needed for the callbacks when a provider is created, which the necessity of is a separate comment. If that gets removed this will too

File: src/coreclr/src/inc/corprof.idl
Comment (davmason) [reply src/coreclr/src/inc/corprof.idl:748]: 2 is DBNULL which EventPipe doesn't support, it felt like a trap to include it

File: src/coreclr/src/inc/corprof.idl
Comment (davmason) [reply src/coreclr/src/inc/corprof.idl:2549]: I initially added it because I wanted to test my sample profiler that parsed eventpipe metadata. Then I could start a session and just add every provider that existed and try to parse every event.

I left it because I can imagine profiler developers using it for finding out what providers are available. I don't have a production scenario in mind. I do think it matches the existing ICorProfiler philosophy of letting the profiler know when something is created.

File: src/coreclr/src/inc/corprof.idl
Comment (davmason) [reply src/coreclr/src/inc/corprof.idl:4125]: We don't expose a mechanism over the pipe, but internally we have the functionality already. When we do rundown we add the rundown provider to the session.

Just like EventPipeProviderCreated above, this was added for testing (to be able to subscribe programatically to all events). But, I can imagine scenarios where it might be easier to add providers rather than starting a new session.

Comment (noahfalk) [reply src/coreclr/src/inc/corprof.idl:4125]: Yeah I think being able to add (and for that matter change/remove) providers is goodness, I was just surprised to see it in the profiler first : )
I'm happy to keep it as long we can deliver the notification outside of the EventPipe lock.

File: src/coreclr/src/inc/corprof.idl
Comment (noahfalk) [reply src/coreclr/src/inc/corprof.idl:2549]: Sounds good to me (as long as we can resolve the lock issue) 👍

**Comment by mjsabby** [comment]
Excited to try this out. @davmason you know if there is anything blocking from merging this in?

**Comment by davmason** [comment]
> 
> 
> Excited to try this out. @davmason you know if there is anything blocking from merging this in?

Noah pointed out some things that need addressed, none of them are fundamental flaws though. I've been working on this in between things so I haven't had time to address them yet. I'm hoping to get back on this soon.

File: src/coreclr/src/inc/corprof.idl
Comment (davmason) [reply src/coreclr/src/inc/corprof.idl:4122]: Looking at this logic, I think there probably is a few bugs here. Nobody every specified the enableSampleProfiler flag in EventPipe::Enable, it was a default arg with value of true. We were relying on SampleProfiler to check if the sampling event was enabled in SampleProfiler::Enable and bail early if it did.

The only place that ever checked for the presence of it was in the enable EventPipe via the environment variable path, but the existing delayed sampling initialization would only increase the SampleProfiler's ref count by 1 no matter how many sessions requested it before the runtime was far enough along.

For my refactoring I'm making EventPipe check if the sample provider is part of the provider list, and if it is unconditionally call SampleProvider::Enable. Then inside of SampleProvider it will delay sampling until the FinishInitialize is called, that way we can have an accurate refcount.

File: src/coreclr/src/vm/eventpipeconfiguration.cpp
Comment (davmason) [reply src/coreclr/src/vm/eventpipeconfiguration.cpp:165]: Turns out, trivial to make the callback outside the lock 😄

**Comment by davmason** [comment]
@noahfalk I believe at this point I have addressed all of your feedback. I pushed changes to fix the SuspendWriteEvent issues, move the profiler callback outside of the lock, and removed the enableCPUSampling argument from EventPipe::Enable.

File: src/coreclr/src/vm/eventpipesession.cpp
Comment (noahfalk) [inline src/coreclr/src/vm/eventpipesession.cpp:582]: I think the logic needs a small adjustment. Imagine:

1. Event writer thread: executed up to but not including [this line](https://github.com/dotnet/runtime/blob/7a8da1a706bd011ec80e7cab28dd0ad38cf54cc3/src/coreclr/src/vm/eventpipe.cpp#L800). 
2. Disable() thread preempts the writer, and executes down to line 586 because no threads report a write in progress.
3. Event writer thread peempts Disable() thread: Executes [Line 802](https://github.com/dotnet/runtime/blob/7a8da1a706bd011ec80e7cab28dd0ad38cf54cc3/src/coreclr/src/vm/eventpipe.cpp#L802) loading a non-null session pointer
4. Disable() thread preempts writer: executes the remainder of EventPipe::Disable, deletes session
5. Event writer thread preempts Disable() thread: [Calls EventPipeSession::WriteEventBuffered()](https://github.com/dotnet/runtime/blob/7a8da1a706bd011ec80e7cab28dd0ad38cf54cc3/src/coreclr/src/vm/eventpipe.cpp#L807) with deleted session pointer, AVs 

The issue appears to be [storing a null pointer into the session array](https://github.com/dotnet/runtime/blob/5f87367da4b55f86410e1e820bd6a4b6439c2a16/src/coreclr/src/vm/eventpipe.cpp#L552) occurs too late. The null store must happen prior to this scan.

After that I think the algorithm is sound:
Writer thread does:
1. [Write InProgress bit = true](https://github.com/dotnet/runtime/blob/7a8da1a706bd011ec80e7cab28dd0ad38cf54cc3/src/coreclr/src/vm/eventpipe.cpp#L800)
2. [Read session pointer](https://github.com/dotnet/runtime/blob/7a8da1a706bd011ec80e7cab28dd0ad38cf54cc3/src/coreclr/src/vm/eventpipe.cpp#L802)

Disable thread does:
1. Write session pointer = null
2. Read InProgress bit (579 above)

In any possible interleaving at least one of the writes must occur before the read on the other thread. If Disable() observes the InProgress bit it waits at least long enough for the write to complete (possibly longer, potentially even allowing multiple new event writes to occur, but the same logic can be applied recursively) so we don't delete the pointer while it is in use. If WriteEventInternal() observes the null session pointer then it bails out without invoking WriteEventBuffered()

**Review by noahfalk — COMMENTED** [review]
I'm still reviewing @davmason but getting late so stopping for the night and I'll keep going tomorrow. So far so good : D

File: src/coreclr/src/vm/eventpipesession.cpp
Comment (davmason) [reply src/coreclr/src/vm/eventpipesession.cpp:422]: I swear I remember adding this, but I either meant to and forgot or accidentally backed it out

File: src/coreclr/src/vm/eventpipesession.cpp
Comment (davmason) [reply src/coreclr/src/vm/eventpipesession.cpp:586]: Thanks Noah. I just pushed changes to move the NULL store to before the call to pSession->SuspendWriteEvent

File: src/coreclr/src/vm/eventpipesession.cpp
Comment (noahfalk) [inline src/coreclr/src/vm/eventpipesession.cpp:571]: This is true, though it feels like an odd thing to assert. Asserting that the s_pSession array slot is null seems like a more direct way to validate that the session will not be writing any more events?

**Review by noahfalk — APPROVED** [review]
I found a couple more issues but nothing that should be too hard to deal with. There are also some old comments like the filterData one that still need to be looked at. But overall, looking good!

File: src/coreclr/src/vm/eventpipesession.cpp
Comment (davmason) [reply src/coreclr/src/vm/eventpipesession.cpp:571]: I just got tunnel visioned. I originally had !IsEnabled(), but the rundown provider was sometimes there so I created this instead of rethinking the approach

File: src/coreclr/src/inc/corprof.idl
Comment (davmason) [reply src/coreclr/src/inc/corprof.idl:787]: From what I can see the format is "key=value;key2=value2;" with optional quotes to escape '=' and ';'. I added a comment to corprof.idl to indicate this.

Comment (noahfalk) [reply src/coreclr/src/inc/corprof.idl:787]: I haven't looked into the implementation required to change it, but just wanted to raise that as a brand new interface we have the opportunity to do so. My guess is that folks using the API wouldn't say escaped string encodings are ideal, but also not a huge deal that blocks their progress. So entirely your call if the implementation cost looks like it would be worth making a change. My recollection is that ETW uses a different encoding where items are separated by embedded null characters with a double-null to indicate the end - no escape characters needed. Or an object model style API could use an array of structs with each struct having key and value fields.

File: src/coreclr/src/vm/eventpipesession.cpp
Comment (noahfalk) [inline src/coreclr/src/vm/eventpipesession.cpp:546]: I meant to comment and forgot, we shouldn't spin waiting for writes to complete inside this lock. WriteEvent() isn't guaranteed to be fast in all cases and all thread creation/destruction in the process has to block on this lock. I'd recommend copying the set of currently active threads into a new snapshot of the list, then iterate over the snapshot outside the lock to wait for in progress writes to complete.

File: src/coreclr/src/vm/eventpipesession.cpp
Comment (noahfalk) [inline src/coreclr/src/vm/eventpipesession.cpp:562]: This comment is no longer accurate, in progress writes finished above. All this does now is convert the thread write buffers into the read-only state ensuring they will be flushed in WriteAllBuffersToFile() later.

**Comment by davmason** [comment]
Filed #38492 to track documentation.

@mjsabby fyi the first preview this will be in is preview 8.

**Comment by k15tfu** [comment]
FYI: https://github.com/dotnet/runtime/pull/41791

**Pull request merged**
